modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/LogicalRelImplementor.java

```diff
@@ -1104,4 +1123,38 @@ private static BinaryTupleSchema fromTableDescriptor(TableDescriptor descriptor)
 
         return BinaryTupleSchema.create(elements);
     }
+
+    private static boolean canFuseProjectionInto(RelNode rel) {
+        if (rel instanceof Join) {
+            Join join = (Join) rel;
+
+            return joinNeedsProjection(join);
+        }
+
+        return false;
+    }
+
+    private static boolean joinNeedsProjection(Join join) {
+        return join.getJoinType() == JoinRelType.INNER
+                || join.getJoinType() == JoinRelType.LEFT
+                || join.getJoinType() == JoinRelType.FULL
+                || join.getJoinType() == JoinRelType.RIGHT;
+    }
+
+    private @Nullable SqlJoinProjection<RowT> createJoinProjection(Join rel, RelDataType outType) {
+        SqlJoinProjection<RowT> joinProjection = null;
+        if (projectionToFuse != null) {
+            assert joinNeedsProjection(rel);
+
+            joinProjection = expressionFactory.joinProject(projectionToFuse, outType);
+
+            projectionToFuse = null;
```

Comment:
Let's have a test that shows projectionToFuse is never used for inappropriate join.
E.g. if someone remove this line or forget to call this method right after `projectionToFuse` field was set.

Reply:
I've updated `joinWithProjectionOnTopReturnsValidaResults` to fail if someone will forget to nullify `projectionToFuse`